File: scripts/session-stats-dashboard.py
#!/usr/bin/env python3
"""
Session Stats Dashboard - Shows session statistics at session end
Hook: SessionEnd

Displays:
- Session duration
- Skills invoked
- Tools used
- Files modified
- Summary of work done
"""

import json
import sys
from pathlib import Path
from datetime import datetime
from collections import Counter

def parse_transcript(transcript_path: Path) -> dict:
    """Parse session transcript and extract statistics."""
    stats = {
        "start_time": None,
        "end_time": None,
        "skills_used": Counter(),
        "tools_used": Counter(),
        "files_modified": set(),
        "user_prompts": 0,
        "assistant_responses": 0,
    }

    if not transcript_path.exists():
        return stats

    try:
        with open(transcript_path, 'r', encoding='utf-8') as f:
            for line in f:
                try:
                    entry = json.loads(line.strip())

                    # Track timestamps
                    if 'timestamp' in entry:
                        ts = datetime.fromisoformat(entry['timestamp'].replace('Z', '+00:00'))
                        if stats["start_time"] is None:
                            stats["start_time"] = ts
                        stats["end_time"] = ts

                    # Count user prompts
                    if entry.get('role') == 'user':
                        stats["user_prompts"] += 1

                    # Count assistant responses
                    elif entry.get('role') == 'assistant':
                        stats["assistant_responses"] += 1

                        # Extract tool uses from content
                        content = entry.get('content', [])
                        if isinstance(content, list):
                            for item in content:
                                if isinstance(item, dict):
                                    # Skill invocations
                                    if item.get('type') == 'tool_use' and item.get('name') == 'Skill':
                                        skill = item.get('input', {}).get('skill', 'unknown')
                                        stats["skills_used"][skill] += 1

                                    # Other tools
                                    elif item.get('type') == 'tool_use':
                                        tool = item.get('name', 'unknown')
                                        stats["tools_used"][tool] += 1

                                        # Extract file paths
                                        tool_input = item.get('input', {})
                                        if 'file_path' in tool_input:
                                            stats["files_modified"].add(tool_input['file_path'])

                except json.JSONDecodeError:
                    continue
    except Exception:
        pass

    return stats


def format_duration(start: datetime, end: datetime) -> str:
    """Format duration as human-readable string."""
    if not start or not end:
        return "unknown"

    delta = end - start
    hours = int(delta.total_seconds() // 3600)
    minutes = int((delta.total_seconds() % 3600) // 60)
    seconds = int(delta.total_seconds() % 60)

    if hours > 0:
        return f"{hours}h {minutes}m {seconds}s"
    elif minutes > 0:
        return f"{minutes}m {seconds}s"
    else:
        return f"{seconds}s"


def display_dashboard(stats: dict):
    """Display session statistics dashboard."""
    print("\n" + "=" * 60)
    print("📊 SESSION STATISTICS")
    print("=" * 60)

    # Duration
    if stats["start_time"] and stats["end_time"]:
        duration = format_duration(stats["start_time"], stats["end_time"])
        print(f"\n⏱️  Duration: {duration}")

    # Interaction count
    print(f"💬 Exchanges: {stats['user_prompts']} prompts, {stats['assistant_responses']} responses")

    # Skills used
    if stats["skills_used"]:
        print(f"\n🔧 Skills used ({len(stats['skills_used'])}):")
        for skill, count in stats["skills_used"].most_common(5):
            print(f"   • {skill} ({count}x)")
        if len(stats["skills_used"]) > 5:
            print(f"   ... and {len(stats['skills_used']) - 5} more")

    # Top tools
    if stats["tools_used"]:
        print(f"\n🛠️  Top tools:")
        for tool, count in stats["tools_used"].most_common(5):
            print(f"   • {tool} ({count}x)")

    # Files modified
    if stats["files_modified"]:
        file_count = len(stats["files_modified"])
        print(f"\n📝 Files modified: {file_count}")
        if file_count <= 5:
            for f in sorted(stats["files_modified"]):
                print(f"   • {f}")
        else:
            for f in sorted(stats["files_modified"])[:3]:
                print(f"   • {f}")
            print(f"   ... and {file_count - 3} more")

    print("\n" + "=" * 60)
    print("Session saved to memory. Use /system:semantic-memory-search to recall.")
    print("=" * 60 + "\n")


def main():
    try:
        # Read hook input
        input_data = json.load(sys.stdin)
        transcript_path = Path(input_data.get('transcript_path', ''))

        if not transcript_path.exists():
            sys.exit(0)

        # Parse and display stats
        stats = parse_transcript(transcript_path)
        display_dashboard(stats)

        sys.exit(0)
    except Exception as e:
        # Silent fail - don't break session end
        print(f"[stats] Error: {e}", file=sys.stderr)
        sys.exit(0)


if __name__ == '__main__':
    main()

File: scripts/session-stats-dashboard.js
#!/usr/bin/env node
/**
 * Session Stats Dashboard - Shows session statistics at session end
 * Hook: SessionEnd
 *
 * Displays:
 * - Session duration
 * - Skills invoked
 * - Tools used
 * - Files modified
 * - Summary of work done
 */

const fs = require('fs');
const path = require('path');

// Add lib directory to path for imports
const libDir = path.join(__dirname, 'lib');
const debugLogger = require(path.join(libDir, 'debug-logger.js'));

const { logHookStart, logHookEnd, logDebug } = debugLogger;

function parseTranscript(transcriptPath) {
    const stats = {
        start_time: null,
        end_time: null,
        skills_used: new Map(),
        tools_used: new Map(),
        files_modified: new Set(),
        user_prompts: 0,
        assistant_responses: 0
    };

    if (!fs.existsSync(transcriptPath)) {
        return stats;
    }

    try {
        const content = fs.readFileSync(transcriptPath, 'utf-8');
        const lines = content.split('\n').filter(l => l.trim());

        for (const line of lines) {
            try {
                const entry = JSON.parse(line);

                // Track timestamps
                if (entry.timestamp) {
                    const ts = new Date(entry.timestamp);
                    if (!stats.start_time) {
                        stats.start_time = ts;
                    }
                    stats.end_time = ts;
                }

                // Count user prompts
                if (entry.role === 'user') {
                    stats.user_prompts++;
                }

                // Count assistant responses
                else if (entry.role === 'assistant') {
                    stats.assistant_responses++;

                    // Extract tool uses from content
                    const content = entry.content || [];
                    if (Array.isArray(content)) {
                        for (const item of content) {
                            if (typeof item === 'object' && item !== null) {
                                // Skill invocations
                                if (item.type === 'tool_use' && item.name === 'Skill') {
                                    const skill = (item.input || {}).skill || 'unknown';
                                    stats.skills_used.set(skill, (stats.skills_used.get(skill) || 0) + 1);
                                }

                                // Other tools
                                else if (item.type === 'tool_use') {
                                    const tool = item.name || 'unknown';
                                    stats.tools_used.set(tool, (stats.tools_used.get(tool) || 0) + 1);

                                    // Extract file paths
                                    const toolInput = item.input || {};
                                    if (toolInput.file_path) {
                                        stats.files_modified.add(toolInput.file_path);
                                    }
                                }
                            }
                        }
                    }
                }
            } catch (e) {
                // Skip malformed lines
                continue;
            }
        }
    } catch (e) {
        // Silent fail
    }

    return stats;
}

function formatDuration(start, end) {
    if (!start || !end) {
        return 'unknown';
    }

    const delta = (end - start) / 1000; // seconds
    const hours = Math.floor(delta / 3600);
    const minutes = Math.floor((delta % 3600) / 60);
    const seconds = Math.floor(delta % 60);

    if (hours > 0) {
        return `${hours}h ${minutes}m ${seconds}s`;
    } else if (minutes > 0) {
        return `${minutes}m ${seconds}s`;
    } else {
        return `${seconds}s`;
    }
}

function displayDashboard(stats) {
    console.log('\n' + '='.repeat(60));
    console.log('📊 SESSION STATISTICS');
    console.log('='.repeat(60));

    // Duration
    if (stats.start_time && stats.end_time) {
        const duration = formatDuration(stats.start_time, stats.end_time);
        console.log(`\n⏱️  Duration: ${duration}`);
    }

    // Interaction count
    console.log(`💬 Exchanges: ${stats.user_prompts} prompts, ${stats.assistant_responses} responses`);

    // Skills used
    if (stats.skills_used.size > 0) {
        console.log(`\n🔧 Skills used (${stats.skills_used.size}):`);
        const sortedSkills = Array.from(stats.skills_used.entries())
            .sort((a, b) => b[1] - a[1])
            .slice(0, 5);

        for (const [skill, count] of sortedSkills) {
            console.log(`   • ${skill} (${count}x)`);
        }
        if (stats.skills_used.size > 5) {
            console.log(`   ... and ${stats.skills_used.size - 5} more`);
        }
    }

    // Top tools
    if (stats.tools_used.size > 0) {
        console.log(`\n🛠️  Top tools:`);
        const sortedTools = Array.from(stats.tools_used.entries())
            .sort((a, b) => b[1] - a[1])
            .slice(0, 5);

        for (const [tool, count] of sortedTools) {
            console.log(`   • ${tool} (${count}x)`);
        }
    }

    // Files modified
    if (stats.files_modified.size > 0) {
        const fileCount = stats.files_modified.size;
        console.log(`\n📝 Files modified: ${fileCount}`);
        const sortedFiles = Array.from(stats.files_modified).sort();

        if (fileCount <= 5) {
            for (const f of sortedFiles) {
                console.log(`   • ${f}`);
            }
        } else {
            for (const f of sortedFiles.slice(0, 3)) {
                console.log(`   • ${f}`);
            }
            console.log(`   ... and ${fileCount - 3} more`);
        }
    }

    console.log('\n' + '='.repeat(60));
    console.log('Session saved to memory. Use /system:semantic-memory-search to recall.');
    console.log('='.repeat(60) + '\n');
}

/* istanbul ignore next */
function main() {
    logHookStart('SessionEnd', 'session-stats-dashboard.js');

    let inputData = '';
    process.stdin.setEncoding('utf8');

    process.stdin.on('readable', () => {
        let chunk;
        while ((chunk = process.stdin.read()) !== null) {
            inputData += chunk;
        }
    });

    process.stdin.on('end', () => {
        try {
            const data = JSON.parse(inputData);
            const transcriptPath = data.transcript_path || '';

            logDebug('SessionEnd', 'session-stats-dashboard.js', `Transcript: ${transcriptPath}`, 'INFO');

            if (!fs.existsSync(transcriptPath)) {
                logDebug('SessionEnd', 'session-stats-dashboard.js', 'Transcript not found', 'SKIP');
                logHookEnd('SessionEnd', 'session-stats-dashboard.js', true);
                process.exit(0);
            }

            // Parse and display stats
            logDebug('SessionEnd', 'session-stats-dashboard.js', 'Parsing transcript...', 'INFO');
            const stats = parseTranscript(transcriptPath);

            logDebug('SessionEnd', 'session-stats-dashboard.js', `Stats: ${stats.user_prompts} prompts`, 'INFO');
            logDebug('SessionEnd', 'session-stats-dashboard.js', 'Displaying dashboard', 'OUTPUT');

            displayDashboard(stats);

            logHookEnd('SessionEnd', 'session-stats-dashboard.js', true);
            process.exit(0);

        } catch (e) {
            logDebug('SessionEnd', 'session-stats-dashboard.js', `ERROR: ${e.message}`, 'ERROR');
            logHookEnd('SessionEnd', 'session-stats-dashboard.js', false);
            // Silent fail - don't break session end
            console.error(`[stats] Error: ${e.message}`);
            process.exit(0);
        }
    });
}

// Export for testing
/* istanbul ignore else */
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        parseTranscript,
        formatDuration,
        displayDashboard
    };
}

// Run main if executed directly
/* istanbul ignore if */
if (require.main === module) {
    main();
}
